packages/router/router.ts

```diff
@@ -815,11 +818,26 @@ export function createRouter(init: RouterInit): Router {
       ...init.history.encodeLocation(location),
     };
 
-    let historyAction =
-      (opts && opts.replace) === true ||
-      (submission != null && isMutationMethod(submission.formMethod))
-        ? HistoryAction.Replace
-        : HistoryAction.Push;
+    let historyAction = HistoryAction.Push;
+
+    if (opts && opts?.replace === true) {
+      historyAction = HistoryAction.Replace;
+    } else if (opts && opts.replace === false) {
+      // no-op
+    } else if (
+      submission != null &&
+      isMutationMethod(submission.formMethod) &&
+      parsePath(submission.formAction).pathname === state.location.pathname
+    ) {
+      // TODO check this with Ryan
+
+      // By default on submissions to the current location we REPLACE so that
+      // users don't have to double-click the back button to get to the prior
+      // location.  If the user redirects from the action/loader this will be
+      // ignored and the redirect will be a PUSH
+      historyAction = HistoryAction.Replace;
+    }
```

Comment:
@ryanflorence This fixes 2 things from the prior approach:

* Explicit `replace:false` was respected if the action returned a redirect, but was getting lost if it didn't
* Our "default to replace on submission" logic is now limited to same-path submissions, since you've already gone to the new location we don't need to replace the location you submitted from to avoid the double-back-click
  * `/`, link to `/a`, submit to `/a` -> history stack is `[/, /a]` - back button takes you to `/`
    * No changes here
  * `/`, link to `/a`, submit to `/b` -> history stack is `[/, /a, /b]` - back button takes you to `/a`
    * Previously this would have replaced `/a` and left you with `[/, /b]` and a back button would go back to `/`